File: map_filter_reduce/java/map_filter_reduce.java
import java.util.Collection;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.*;

class Main {
    public static void main(String[] args) {
        List<Integer> myList = new ArrayList<Integer>();
        for (int i = 1; i < 20; i++) {
            myList.add(i);
        }
        
        // Convert it into a Stream
        Stream<Integer> myStream = myList.stream();

        // Square elements
        // Take only ones less than 50
        // Multiply remaining together
        int result = myStream
            .map(elem -> elem * elem)
            .filter(elem -> elem < 50)
            .reduce(1, (acc, elem) -> acc * elem);

        System.out.println(result);
    }
}

File: map_filter_reduce/lua/map_filter_reduce.lua
local function map(func, tab)
    local new_tab = {}
    for i,elem in pairs(tab) do
        new_tab[i] = func(elem)
    end
    return new_tab
end
local function filter(func, tab)
    local new_tab = {}
    for i,elem in pairs(tab) do
        if func(elem) then
            new_tab[i] = elem
        end
    end
    return new_tab
end
local function reduce(identity, func, tab)
    local result = identity
    for _,v in pairs(tab) do
        result = func(result, v)
    end
    return result
end
local tab = {}
for i=1,20 do tab[#tab + 1] = i end
-- Square elements
-- Take only ones less than 50
-- Multiply remaining together
Result = reduce(
    1,
    (function(acc, elem) return elem * acc end),
    filter(
        (function(elem) return elem < 50 end),
        map(
            (function(elem) return elem*elem end), 
            tab)
    )
)
print(Result)